Use struct for `NodeDurations` in both `ALAP` and `ASAP` Scheduling Passes

raynelfss opened the pull request:
<!--
⚠️  If you do not respect this template, your pull request will be closed.
⚠️  Your pull request title should be short detailed and understandable for all.
⚠️  Also, please add a release note file using reno if the change needs to be documented in the release notes.
⚠️  If your pull request fixes an open issue, please link to the issue. Use "Fixes #XXXX" if this PR *fully* closes the issue XXXX.  
☢️  If you used an AI tool to code this PR, add "AI tool used: <Name and version of the tool>". For example, "AI tool used: Microsoft Copilot Chat with GPT-5". Failing to disclose the use of AI tools may result in the PR being closed without further review.  


- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary
The following commits introduce a PyO3 class, `NodeDurations`, that represents a mapping between `DAGOpNode` instances and their duration values in either units of `dt` or `seconds`.

The main purpose of this structure is to avoid the costly conversion from `PyDict` to `HashMap` and back in each instance.

The following PR also restructures the directories for scheduling passes to allow this struct to live in its own file under a common directory.


### Details and comments

raynelfss committed:
Initial: Reorganize schedule analysis passes
- Add initial ``NodeDurations`` struct.

raynelfss committed:
Fix: Use NodeDurations struct on passes
This commit is incomplete and most features will stop working.

raynelfss committed:
Add: Use `NodeDurations` in python.
- Add magic methods to `PyNodeDurations`.

raynelfss committed:
Fix: Use indices to get nodes from Dynamical Decoupling passes
- Add `items` method.

raynelfss committed:
Lint: Formatting

raynelfss committed:
Fix: Keep original dag nodes and give path to getting them from scratch.
- Add an extra field to `PyNodeDurations` to keep mapping between indices and original dag nodes.
- Add methods to update `NodeDurations` using mappings with a subset of the current nodes or new mappings as long as the user has the original dag.
- Reverted some changes inside of Context Aware DD.

raynelfss commented:
Does this PR need a release note? Although it is a very internal change, the property slot that stores it is mentioned in [documentation](https://quantum.cloud.ibm.com/docs/en/api/qiskit/transpiler#scheduling-stage), but its type is never mentioned explicitly.

raynelfss committed:
Format: Elide lifetimes.

alexanderivrii commented on crates/transpiler/src/passes/schedule_analysis/mod.rs:
```diff
@@ -0,0 +1,369 @@
+
+/// A mapping between a DAG's [NodeIndex] and its duration values.
+///
+/// A duration may be in units of Dt, represented by a `u32`, or Seconds,
```
```suggestion
/// A duration may be in units of Dt, represented by a `u64`, or Seconds,
```

alexanderivrii commented on crates/transpiler/src/passes/schedule_analysis/mod.rs:
```diff
@@ -0,0 +1,369 @@
+                ))
+            }
+        } else {
+            Ok(PyNodeDurations {
+                inner: NodeDurations::Dt(Default::default()),
+                nodes_mapping: Default::default(),
+            })
```
To confirm: for an empty dict, the default durations are of type `dt`?

raynelfss replied:
Yes, that's although it is mostly an assumption but in most cases we'd deal with cases of `dt`. Do you think we should consider `s` as the default here?

alexanderivrii commented on crates/transpiler/src/passes/schedule_analysis/mod.rs:
```diff
@@ -0,0 +1,369 @@
+                    nodes_mapping: idx_to_node,
+                })
+            } else if key.extract::<f64>().is_ok() {
+                // All durations are of type u64
```
```suggestion
                // All durations are of type f64
```

alexanderivrii commented on crates/transpiler/src/passes/schedule_analysis/mod.rs:
```diff
@@ -0,0 +1,369 @@
+        if self.len() != other.len()? {
+            return Ok(false);
+        }
+        if let Ok(other_nodes) = other.cast::<PyNodeDurations>() {
+            Ok(other_nodes.borrow().eq(self))
+        } else if let Ok(as_dict) = other.cast::<PyDict>() {
+            let as_node_durations = PyNodeDurations::new(as_dict)?;
+            Ok(as_node_durations.inner == self.inner)
+        } else {
```
Do we have Python tests for each of these branches?

raynelfss replied:
We do indirectly in some tests that compare `NodeDurations` to a `dict` instance, for example:
https://github.com/Qiskit/qiskit/blob/676863eed5d57a55e27f4b3e35e86e520e565967/test/python/transpiler/test_scheduling_padding_pass.py#L144

Although this is a very barebones test it allows me not to have to modify all test files and allow for an easier comparison.

alexanderivrii commented on crates/transpiler/src/passes/schedule_analysis/mod.rs:
```diff
@@ -0,0 +1,369 @@
+            return Ok(false);
+        }
+        if let Ok(other_nodes) = other.cast::<PyNodeDurations>() {
+            Ok(other_nodes.borrow().eq(self))
```
Does this check both the equality of `nodes_mapping` and of `inner`? Since `inner` is of type `IndexMap`, does the equality depend on the insertion order?

raynelfss replied:
In this case it should only compare `inner`. 

The reason mostly being that you can't really compare the Python `DAGOpNodes` from rust without a `py` token. 

Now, you may think that because of that, we shouldn't be able to call `::eq`. Furthermore, if you noticed in the definition, `PyNodeDurations` doesn't implement `Eq` or `PartialEq`.  However, the inner `NodeDurations` object does and `PyNodeDurations` implements `Deref`-ing into a `NodeDurations` object, allowing it to access the inner's `::eq` method and avoiding comparing the `DagOpNodes`.

alexanderivrii replied:
Thanks for the explanation!

alexanderivrii commented on crates/transpiler/src/passes/schedule_analysis/mod.rs:
```diff
@@ -0,0 +1,369 @@
+            Ok(other_nodes.borrow().eq(self))
+        } else if let Ok(as_dict) = other.cast::<PyDict>() {
+            let as_node_durations = PyNodeDurations::new(as_dict)?;
+            Ok(as_node_durations.inner == self.inner)
```
Here we only check the equality of `inner`, not of `nodes_mapping`, correct? Same question on the insertion order.

raynelfss replied:
Correct! The insertion order here will be tracked by `NodeDurations` which is why it uses an `IndexMap` instead of a `HashMap` like the op-nodes mapping here does.

Update: Insertion order is no longer relevant after 83127ae

alexanderivrii commented on crates/transpiler/src/passes/schedule_analysis/mod.rs:
```diff
@@ -0,0 +1,369 @@
+
+        match &mut self.inner {
+            NodeDurations::Dt(hash_map) => hash_map.shift_remove(&idx).into_bound_py_any(py),
+            NodeDurations::Seconds(hash_map) => hash_map.shift_remove(&idx).into_bound_py_any(py),
```
I am wondering why we need to use `shift_remove` and not `swap_remove` here.

raynelfss replied:
This is a question that I had too, I did this to preserve the insertion order since some passes would complaint if I didn't include it. Should the insertion order matter at all?

raynelfss replied:
Removed with the usage of `IndexMap` in 83127ae

alexanderivrii reviewed (commented):
Thanks Ray, I have left a few questions (in particular there is something about the equality of `PyNodeDurations` that I do not quite understand).

raynelfss committed:
Apply suggestions from code review

Co-authored-by: Alexander Ivrii <alexi@il.ibm.com>

raynelfss committed:
Merge branch 'main' into struct-node-start

alexanderivrii reviewed (dismissed):
Ray, thanks for explanations and quick updates. Based on Matthew's comment that node durations should only be used for look-ups by node id, I think that using `HashMap` to represent node durations and iterating over DAG's `Delay` nodes in dynamical decoupling both make sense.

This PR does change the _type_ of `property_set["node_duration"]` from `dict` to `class 'qiskit._accelerate.scheduling.NodeDurations`, however I don't think this poses a problem from the backward compatibility standpoint.

raynelfss committed:
Fix: Add missing magic methods.

raynelfss committed:
Merge branch 'main' into struct-node-start

raynelfss committed:
Fix: Pre-computing error in `__getitem__`
- An oversight allowed an error variant to precompute itself even if the result was `Ok` via `ok_or`. Replaced it with `ok_or_else`.
- Fix formatting issue.

raynelfss commented:
@alexanderivrii I reverted the changes back to using `IndexMap` and we're lucky to have done so as during testing I found a performance regression I had accidentally introduced 😅. The PR should be in a much better state now.

raynelfss committed:
Merge branch 'main' into struct-node-start

alexanderivrii reviewed (commented):
Thanks for the updates, Ray! 

I did not fully understand where the performance regression that you mentioned was coming from, is it from using `HashMap` instead of `IndexMap`?

Overall this looks very nice to me. I believe to _fully_ look a Python `dict`, you should also implement the `__iter__` magic method, what do you think? 

Otherwise, it's ready to merge.

raynelfss commented:
> I did not fully understand where the performance regression that you mentioned was coming from, is it from using `HashMap` instead of `IndexMap`?

Oh, I forgot to explain it but it was part of the commit message of the last commit:

cf08009
```
Fix: Pre-computing error in __getitem__
- An oversight allowed an error variant to precompute itself even if the result was `Ok` via `ok_or`. Replaced it with `ok_or_else`.
...
```

The regression I introduced simply called Python's `__repr__` method on each node regardless of if it was to be used for an error message or not, which caused the performance to worsen greatly.

> I believe to fully look a Python dict, you should also implement the __iter__ magic method, what do you think?

I might be wrong but because the struct is specified to be a `mapping` and it implements the `__getitem__`, `__setitem__` and `__len__` methods, Python provides a default implementation. Or at least that what it's supposed to do. I haven't tested this so allow me to confirm that it is possible.

**Update:** it in fact does not make it iterable...

raynelfss committed:
Add: `__iter__`, `keys` and `values` to `NodeDurations`.

raynelfss commented:
Added iteration methods (`__iter__`, `keys` and `values`) in 95ccf0c

alexanderivrii reviewed (approved):
Thanks for the explanations, and thanks for remembering to also implement `keys` and `values`. After experimenting with the code, the `NodeDurations` data structure now looks like a dict and quacks like a dict, therefore for all practical purposes it's backwards compatible.
